riscv/insns/c_sspopchk_x5.h

```diff
@@ -0,0 +1,32 @@
+reg_t prv = STATE.prv;
+reg_t xSSE = 0;
+
+if (prv == PRV_M)
+  xSSE = 0;
+else if (prv == PRV_S || prv == PRV_HS)
+  xSSE = get_field(STATE.menvcfg->read(), MENVCFG_SSE);
+else if (p->extension_enabled('S'))
+  xSSE = get_field(STATE.senvcfg->read(), SENVCFG_SSE);
+else
+  xSSE = false;
+
+// For VS-mode
+if (prv == PRV_S && STATE.v)
+  xSSE &= get_field(STATE.henvcfg->read(), HENVCFG_SSE);
+
+if (!xSSE) {
+  #include "c_mop_N.h"
+} else {
+  reg_t ssp = STATE.ssp->read();
+  const auto new_sp = ssp + xlen / 8;
+  uint64_t ss_addr;
+
+  if (xlen == 32)
+    ss_addr = MMU.ss_load<uint32_t>(ssp);
+  else
+    ss_addr = MMU.ss_load<uint64_t>(ssp);
+
+  software_check(T0 == ss_addr, SHADOW_STACK_FAULT);
+
```

Comment:
We could avoid duplicating code between 3 flavors of sspopchk. We can create a macro file for zicfiss and put the template for popchk and template for push in that header file and include it here. 
We also dont need
"if (!xSSE) {
   #include "c_mop_N.h"
} else
Since neither push nor popchk write destination (their rd is 0) and compressed also do not write destination., we can simply do:
if (xSSE) {
    // the ss instruction behavior
}

Reply:
Thank you very much for your suggestions.
We can create some zicfiss macro in zicfiss.h.
```
#define check_ssp(reg, ssp) \
  reg_t shadow_return_addr; \
  if (xlen == 32) \
    shadow_return_addr = MMU.ss_load<uint32_t>(ssp); \
  else \
    shadow_return_addr = MMU.ss_load<uint64_t>(ssp); \
  software_check(reg == shadow_return_addr, SHADOW_STACK_FAULT);

#define store_to_ssp(ssp, reg) \
  if (xlen == 32) \
    MMU.ss_store<uint32_t>(ssp, reg); \
  else \
    MMU.ss_store<uint64_t>(ssp, reg);
```